examples/QubitRotation.py

```diff
@@ -0,0 +1,129 @@
+# e^{-i(\phi-\omega)/2}\sin(\theta/2) & e^{i(\phi+\omega)/2}\cos(\theta/2)
+# \end{bmatrix}$
+# 
+# to reach the target state, which is defined to be $|0 \rangle$ in this case.
+
+
+import jax.numpy as jnp
+import numpy as onp
+from qutip import *
+from jax import grad, jit
+from jax.random import normal, PRNGKey
+
+
+def rot(phi, theta, omega):
+    """Returns a rotation matrix describing rotation around Z-Y-Z axis
+    
+    Parameters:
+    ----------
+        phi: int/float 
+             rotation angle for the second rotation around z axis 
+             
+        theta: int/float
+             rotation angle around y axis
+             
+        omega: int/float
+             rotation angle for the first rotation around z axis
+        
+    Returns:
+    --------
+        rm: array[complex]
+             array representing the rotation matrix 
+    """
+    
+    cos = jnp.cos(theta/2)
+    sin = jnp.sin(theta/2)
+
+    return jnp.array(
+    [
+        [jnp.exp(-0.5j*(phi+omega))*cos, -(jnp.exp(0.5j*(phi-omega)))*sin],
+        [jnp.exp(-0.5j*(phi-omega))*sin, jnp.exp(0.5j*(phi+omega))*cos],
+        
+    ]
+    )
+
+
+def cost(phi, theta, omega, ket):
+    r"""Returns the fidelity between the evolved state and :math: `|0 \rangle` state
+    
+    Parameters:
+    ----------
+        phi: int/float 
+             rotation angle for the second rotation around z axis 
+             
+        theta: int/float
+             rotation angle around y axis
+             
+        omega: int/float
+             rotation angle for the first rotation around z axis
+             
+        ket: array[complex] 
+             array representing the ket to be acted upon by by the rotation matrix
+
+    Returns:
+    -------
+        float: 
+        fidelity between the :math:`|0 \rangle` state and the evolved state under rotation    
+    """
+    
+    evolved = jnp.dot(rot(phi, theta, omega), ket)
+    return jnp.absolute(jnp.vdot(evolved.T,basis(2,0).full()))
+
+
+# ## Gradient Descent Implementation 
+# 
+# Here we have our own gradient descent implementation. There are multiple paramters, like `epochs`, `alpha`, etc, that one might tune further in hopes of better convergence. For our purposes, however, we see that our implementation converges well in a little less than 100 steps. 
+
+
+
+#initialize parameters for descent
+epochs = 100
+alpha = 0.1 #learning rate
+tol = 1e-7
+diff = 1 # diff of new and prev weights should be less than diff
+max_iters = 1
+iters = 0
+key = PRNGKey(0) #Generate a PRNG key
+weights = normal(key=key, shape=(3,))
+init_ket = basis(2,1).full()
+der_cost = grad(cost, argnums = [0,1,2])
+
+for epoch in range (epochs):
+    iters = 0
+    diff = 1
+    tol = 1e-7
+    while jnp.all(diff > tol) and iters < max_iters:
+        prev_weights = weights
+        der = jnp.asarray(der_cost(*prev_weights.T,init_ket))
+        weights = weights + alpha*der
+        iters += 1
+        diff = jnp.absolute(weights-prev_weights)
```

Comment:
This block seems to do nothing. If you see the iters, the while loop exists always after each iteration. I understand the convergence criteria that you want to use is that if the weights do not change too much then break out of the loop. However I do not see the necessity of making two blocks here using epochs and the inner loop. You could just use one of the loops and it still works.

Reply:
I made the epochs loop here just in case a user wants to use epochs. Better code reuse